pydoctor/model.py

```diff
@@ -808,15 +822,23 @@ def ensurePackage(self, package_full_name: str) -> _PackageT:
 
     def _introspectThing(self, thing: object, parent: Documentable, parentMod: _ModuleT) -> None:
         for k, v in thing.__dict__.items():
-            if (isinstance(v, (types.BuiltinFunctionType, types.FunctionType))
+            # TODO(ntamas): MethodDescriptorType and ClassMethodDescriptorType are Python 3.7 only.
+            if (isinstance(v, func_types)
                     # In PyPy 7.3.1, functions from extensions are not
-                    # instances of the above abstract types.
-                    or v.__class__.__name__ == 'builtin_function_or_method'):
+                    # instances of the abstract types in func_types
+                    or (hasattr(v, "__class__") and v.__class__.__name__ == 'builtin_function_or_method')):
                 f = self.Function(self, k, parent)
                 f.parentMod = parentMod
                 f.docstring = v.__doc__
                 f.decorators = None
-                f.signature = Signature()
+                try:
+                    f.signature = signature(v)
+                except Exception:
```

Comment:
Can we be more specific than `Exception`? I'm worried about accidentally catching exceptions that are caused by bugs rather than broken input.

Looking at the `inspect` module's source, the exceptions that can be raised are `ValueError`, `TypeError` and `RuntimeError`. I assume `RuntimeError` will only be raised when something is inconsistent within CPython itself, so we shouldn't have to catch that. The same might be true for `TypeError`, but I'm less confident there.

Reply:
I'm not sure whether it's good to tie the set of exceptions that we catch to a particular implementation of `inspect.signature()`. The implementation might change any time and since there is no documentation about the exceptions that `inspect.signature()` may throw, I would say it's better to catch everything. After all, the idea is to parse `v` as a signature, and if it fails for any reason, just fall back to the text signature; we don't want to fail the entire documentation build because of that.

Anyhow, if you prefer to be more restrictive here, just let me know which exceptions I should catch and I'll be happy to change it.

Reply:
It seems like `ValueError` is the most common exception; it gets raised when there is no text signature on a method/function from a C extension as well as when there _is_ a text signature but it cannot be parsed. One can distinguish between the two by looking at the presence of the `__text_signature__` attribute. I'll change my code to catch `ValueError` only and we can extend it later if it turns out that there are legitimate cases when Python raises something else and not a `ValueError`.